### ISSUE TITLE
Nullpointerexception in CompletableFuture AsyncServiceClient

### ISSUE THREAD

**Issue by sujith-mn** [devonfw/devon4j#324]
Got the below error while calling the remote service asynchronously.

2020-12-15 21:05:42.129  INFO 11620 --- [onPool-worker-3] m.s.c.b.c.ServiceClientPerformanceLogger : Invoking service com.example.domain.test1.employeemanagement.service.impl.rest.AsyncEmployeeService#getEmployee[http://localhost:8080/services/rest/employeemanagement/v1/employee/1] took PT47.4093297S (47409329700ns) and succeded with status 302.
java.util.concurrent.ExecutionException: java.lang.NullPointerException
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
	at com.example.domain.test1.employeemanagement.service.impl.rest.EmployeemanagementRestServiceImpl.getEmployee(EmployeemanagementRestServiceImpl.java:49)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:179)
	at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96)
	at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:201)
	at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:104)
	at org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:59)
	at org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:96)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)
	at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121)
	at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:267)
	at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234)
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208)
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160)
	at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:225)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:301)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doGet(AbstractHTTPServlet.java:225)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:645)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:276)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at com.devonfw.module.logging.common.impl.DiagnosticContextFilter.doFilter(DiagnosticContextFilter.java:84)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at com.devonfw.module.logging.common.impl.PerformanceLogFilter.doFilter(PerformanceLogFilter.java:61)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:320)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:126)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:90)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:118)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:158)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.web.filter.CorsFilter.doFilterInternal(CorsFilter.java:92)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.header.HeaderWriterFilter.doHeadersAfter(HeaderWriterFilter.java:92)
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:77)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:215)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:178)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:358)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:271)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:373)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1589)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.NullPointerException
	at com.devonfw.module.service.common.base.client.async.AbstractAsyncServiceClient.logError(AbstractAsyncServiceClient.java:106)
	at com.devonfw.module.httpclient.common.impl.AbstractAsyncServiceHttpClient.handleResponse(AbstractAsyncServiceHttpClient.java:149)
	at com.devonfw.module.httpclient.common.impl.AbstractAsyncServiceHttpClient.lambda$doCall$1(AbstractAsyncServiceHttpClient.java:129)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)

**Comment by hohwille** [comment]
The NPE happens here when logging an error:
https://github.com/devonfw/devon4j/blob/7b1d0ecb100900abe952f84297e04e64547e405d/modules/service/src/main/java/com/devonfw/module/service/common/base/client/async/AbstractAsyncServiceClient.java#L106

This is bad as we lost the original exception that was going to be logged.
Have you been able to debug the error and see which object was actually `null`?
The `context` can not be `null` as then the error would have already ocurred in line 105.
Further, `this.stub` should also not be `null`. I would guess that also `getInvocation()` does not return `null` so my most likely guess is that `getMethod()` returned `null`.
Do we have a clue why?
How could I get the code to assist and take over? Or can you do the analysis and provide the details here?

**Comment by hohwille** [comment]
https://github.com/devonfw/devon4j/blob/7b1d0ecb100900abe952f84297e04e64547e405d/modules/service/src/main/java/com/devonfw/module/service/common/base/client/async/ServiceClientInvocationImpl.java#L40

**Comment by hohwille** [comment]
Method can actually not be null as only created here:
https://github.com/devonfw/devon4j/blob/7b1d0ecb100900abe952f84297e04e64547e405d/modules/service/src/main/java/com/devonfw/module/service/common/base/client/async/ServiceClientStubImpl.java#L76
(see above - method comes from reflection API and has been used before being not null).

**Comment by hohwille** [comment]
So I assume that `this.stub.getInvocation()` is actually `null` as the `invocation` gets reset to `null` after it has been received once:
https://github.com/devonfw/devon4j/blob/7b1d0ecb100900abe952f84297e04e64547e405d/modules/service/src/main/java/com/devonfw/module/service/common/base/client/async/ServiceClientStubImpl.java#L48

**Comment by hohwille** [comment]
So an interesting bug finding that should be fixed. However, we still do not know what the actual bug has been. Can be anything wrong configuration, service failed in implementation, authentication or authorization failed, error on client e.g. whilst unmarshalling data, etc.

**Comment by hohwille** [comment]
Pity is that we use this error handler here:
https://github.com/devonfw/devon4j/blob/7b1d0ecb100900abe952f84297e04e64547e405d/modules/service/src/main/java/com/devonfw/module/service/common/base/client/async/AbstractAsyncServiceClient.java#L73
and just there we have the invocation at hand:
https://github.com/devonfw/devon4j/blob/7b1d0ecb100900abe952f84297e04e64547e405d/modules/service/src/main/java/com/devonfw/module/service/common/base/client/async/AbstractAsyncServiceClient.java#L69

As I am not going to change our error handler API adding the invocation to keep compatibility, the best solution is IMHO to internally keep the last invocation inside the default error handler. As the service client is not be used concurrently this is safe to do. I will create a PR fixing this asap.

**Comment by hohwille** [comment]
OK, PR #325 with the fix created.
This NPE has nothing to do with issue #283 and its implementation as PR #313.
It must have already been present in devon4j `2020.08.001` if I am not mistaken:
https://github.com/devonfw/devon4j/blob/release/2020.08.001/modules/service/src/main/java/com/devonfw/module/service/common/base/client/async/AbstractAsyncServiceClient.java#L105

**Comment by hohwille** [comment]
I can confirm that with this fix of PR #325 I now get the exception logged properly:
```
2020-12-17 11:38:24.568 ERROR 49980 --- [onPool-worker-9] d.m.s.c.b.c.a.AbstractAsyncServiceClient : Failed to invoke service com.example.domain.test1.employeemanagement.service.impl.rest.AsyncEmployeeService#getEmployee at http://localhost:8082/services/rest

net.sf.mmm.util.exception.api.ServiceInvocationFailedException: While invoking the service com.example.domain.test1.employeemanagement.service.impl.rest.AsyncEmployeeService#getEmployee[http://localhost:8082/services/rest/employeemanagement/v1/employee/1] the following error occurred: An unexpected error has occurred! We apologize any inconvenience. Please try again later.. Probably the service is temporary unavailable. Please try again later. If the problem persists contact your system administrator.
4efa8faa-c790-4f67-be6c-f3442a810032:TechnicalError
	at com.devonfw.module.service.common.impl.client.ServiceClientErrorFactoryImpl.create(ServiceClientErrorFactoryImpl.java:30) ~[devon4j-service-2020.12.001.jar:2020.12.001]
	at com.devonfw.module.service.common.base.client.AbstractServiceClientErrorFactory.createException(AbstractServiceClientErrorFactory.java:77) ~[devon4j-service-2020.12.001.jar:2020.12.001]
	at com.devonfw.module.service.common.base.client.AbstractServiceClientErrorFactory.unmarshall(AbstractServiceClientErrorFactory.java:65) ~[devon4j-service-2020.12.001.jar:2020.12.001]
	at com.devonfw.module.httpclient.common.impl.AbstractAsyncServiceHttpClient.createError(AbstractAsyncServiceHttpClient.java:90) ~[devon4j-http-client-2020.12.001.jar:2020.12.001]
	at com.devonfw.module.httpclient.common.impl.AbstractAsyncServiceHttpClient.handleResponse(AbstractAsyncServiceHttpClient.java:141) ~[devon4j-http-client-2020.12.001.jar:2020.12.001]
	at com.devonfw.module.httpclient.common.impl.AbstractAsyncServiceHttpClient.lambda$doCall$1(AbstractAsyncServiceHttpClient.java:129) ~[devon4j-http-client-2020.12.001.jar:2020.12.001]
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177) ~[na:na]
```
